File: sshiny/src/jlo/ioe/Screen.scala
package jlo.ioe

import java.awt.GraphicsEnvironment
import java.awt.DisplayMode
import java.awt.Dimension
import java.awt.Color
import javax.swing.JFrame
import javax.swing.JComponent
import java.awt.BorderLayout
import javax.swing.UIManager
import javax.swing._
import java.awt.AWTEvent
import java.awt.event._
import scala.actors._
import jlo.ioe.ui._
import jlo.ioe.ui.Panel
import jlo.ioe.command._
import scala.List
import javax.swing.border.LineBorder


@serializable
class ScreenState(name : String) {
  var sheets = List[SheetState]()
  var currentSheet : Option[SheetState] = None

  def name = name
  def addSheet(s:SheetState) = sheets = sheets ::: List(s)
  def removeSheet(s:SheetState) = sheets = sheets.remove {e => s==e}
}

// todo: split this out? probably should be in separate file
class Screen(name : String) extends JFrame("Environment: " + name) with CommandInterceptor {
  var sstate = new ScreenState(name)
  val screenDevice = GraphicsEnvironment.getLocalGraphicsEnvironment().getDefaultScreenDevice()
  val top = new Panel() { setBorder(new LineBorder(Color.gray,1)) } // todo: make INFOPANEL!
  val center = new Panel() { setBorder(null) }
  val sheetSelector = new SheetSelector(width())
  var commandOn = false
  var currentSheet : Option[Sheet] = None
  _setup

  def this(s : ScreenState) = { 
    this(s.name)
    sstate = s 
    for (val ss <- sstate.sheets) {
      
    }
  }

  private def _setup = {
    setRootPane(new RootPane)
    setLayeredPane(new LayeredPane)
    setContentPane(new Panel)
    _layout
    setFullscreen(true)
    setVisible(true) 
    enableEvents(AWTEvent.KEY_EVENT_MASK)
    setFocusable(true)
    getContentPane().asInstanceOf[JComponent].grabFocus()
  }

  private def _layout = {
    setLayout(new BorderLayout())
    top.setBackground(Color.black)
    center.setBackground(Color.white)
    //sheetSelector.setBackground(Color.gray)
    add(top, BorderLayout.NORTH)
    add(center, BorderLayout.CENTER)
    add(sheetSelector, BorderLayout.SOUTH)
    getLayeredPane().setLayout(null)
    validate()
    repaint()
  }

  def nextSheet = currentSheet match {
    case Some(s) => currentSheet = Some(sheetSelector.nextSheet)
    case None => {}
  }

  def prevSheet = currentSheet match {
    case Some(s) => currentSheet = Some(sheetSelector.prevSheet)
    case None => {}
  }

  def display(aSheet:Sheet) = {
    val displayer = new Runnable {
      def run : Unit = {
	currentSheet match {
	  case Some(s) => { center.remove(s); s.setVisible(false); }
	  case None => {}
	}
	aSheet.setVisible(true)
	center.add(aSheet)
	center.invalidate()
	validate()
	repaint()
	currentSheet = Some(aSheet)
      }
    }
    if (Thread.currentThread.getName().startsWith("AWT")) {
      displayer.run
    } else {
      SwingUtilities.invokeLater(displayer)
    }
  }

  def newSheet(a:data.DataObject) = {
    a.defaultView.setPreferredSize(getSize())
    // todo: keep track of the sheets?
    val s = new Sheet(this,a)
    sheetSelector.newSheet(s)
    display(s)
  }

  def splitSheet(a:data.DataObject) = {
    sheetSelector.currentSheet.split(a)
  }

  def showCommand = {
    SwingUtilities.invokeLater(new Runnable {
      def run : Unit = {
	CommandInterface.component.setVisible(true)
	center(CommandInterface.component)
	getLayeredPane().add(CommandInterface.component,CommandInterface.level)
	validate()
	repaint()
      }
    })
  }

  def removeCommand = {
    SwingUtilities.invokeLater(new Runnable {
      def run : Unit = {
	CommandInterface.component.setVisible(false)
	getLayeredPane().remove(CommandInterface.component)
	validate()
	repaint()
	getContentPane().requestFocusInWindow
      }
    })
  }
  
  def center(c:JComponent) = {
    val w = (width()-c.getWidth())/2
    val h = (height()-c.getHeight())/2
    c.setLocation(w,h)
  }
  def width() = screenDevice.getDisplayMode().getWidth()
  def height() = screenDevice.getDisplayMode().getHeight()
  
  private def setFullscreen(b : boolean) = {
    val dm = screenDevice.getDisplayMode()
    if (screenDevice.isFullScreenSupported()) {
      setUndecorated(true)
      setResizable(false)
      screenDevice.setFullScreenWindow(this)
      validate();
    } else
      Console.println("full-screen mode unsupported")
  }
   
  override def commandRequested : Unit = if (!commandOn) { commandOn = true; showCommand }
				  else { commandOn = false; removeCommand }
}

File: trunk/jshiny/src/jlo/ioe/Screen.java
package jlo.ioe;

import jlo.ioe.data.DataObject;
import jlo.ioe.ui.LayeredPane;
import jlo.ioe.ui.Panel;
import jlo.ioe.ui.RootPane;
import jlo.ioe.util.F;
import jlo.ioe.util.Opt;

import javax.swing.JComponent;
import javax.swing.JFrame;
import javax.swing.SwingUtilities;
import javax.swing.border.LineBorder;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.DisplayMode;
import java.awt.GraphicsDevice;
import java.awt.GraphicsEnvironment;
import java.io.Serializable;
import java.util.LinkedList;
import java.util.List;

/**
 * Copyright © 2007 imaginaryday.com (jlo)<br>
 * User: jlowens<br>
 * Date: Apr 7, 2007<br>
 * Time: 2:17:44 PM<br>
 */
public class Screen extends JFrame {
	transient GraphicsDevice screenDevice =
			GraphicsEnvironment.getLocalGraphicsEnvironment().getDefaultScreenDevice();
	Panel top = new Panel() {{ setBorder(new LineBorder(Color.gray,1)); }};
	Panel center = new Panel() {{ setBorder(null); }};
	SheetSelector sheetSelector = new SheetSelector(width());
	boolean commandOn = false;
	ScreenState sstate;

	public Screen(String name) {
		sstate = new ScreenState(name);
		setup();
	}

	private void setup() {
		setRootPane(new RootPane());
		setLayeredPane(new LayeredPane());
		setContentPane(new Panel());
		_layout();
		setFullscreen(true);
		setVisible(true);
		setFocusable(true);
		((JComponent)getContentPane()).grabFocus();
	}

	private void setFullscreen(boolean b) {
		DisplayMode dm = screenDevice.getDisplayMode();
		if (screenDevice.isFullScreenSupported()) {
			setUndecorated(true);
			setResizable(false);
			screenDevice.setFullScreenWindow(this);
			validate();
		} else {
			System.out.println("full-screen mode unsupported");
		}
	}

	private void _layout() {
		setLayout(new BorderLayout());
		top.setBackground(Color.black);
		center.setBackground(Color.white);
		add(top, BorderLayout.NORTH);
		add(center, BorderLayout.CENTER);
		add(sheetSelector, BorderLayout.SOUTH);
		getLayeredPane().setLayout(null);
		validate();
		repaint();
	}

	public Sheet newSheet(DataObject o) {
		return null;
	}

	int width() { return screenDevice.getDisplayMode().getWidth(); }
	int height() { return screenDevice.getDisplayMode().getHeight(); }
	void moveToCenter(JComponent c) {
		int w = (width()-c.getWidth())/2;
		int h = (height()-c.getHeight())/2;
		c.setLocation(w,h);
	}

	public void splitSheet(DataObject o) {

	}

	public void nextSheet() {
		sstate.currentSheet.match(
				new F.lambda1<Object,Sheet>(){protected Object code(Sheet p) {
					sstate.currentSheet = Opt.some(sheetSelector.nextSheet()); return null;
				}},
				new F.lambda1<Object, Sheet>(){protected Object code(Sheet p) {
					return null;
				}});
	}

	public void prevSheet() {
		sstate.currentSheet.match(
				new F.lambda1<Object,Sheet>(){protected Object code(Sheet p) {
					sstate.currentSheet = Opt.some(sheetSelector.prevSheet()); return null;
				}},
				new F.lambda1<Object, Sheet>(){protected Object code(Sheet p) {
					return null;
				}});
	}

	public void commandRequested() {
		if (!commandOn) {
			commandOn = true; showCommand();
		} else {
			commandOn = false; removeCommand();
		}
	}

	private void removeCommand() {
		SwingUtilities.invokeLater(new Runnable() {
			public void run() {
				
			}
		});
	}

	private void showCommand() {

	}

	public void display(Sheet sheet) {

	}

	public static class ScreenState implements Serializable {
		public List<Sheet> sheets = new LinkedList<Sheet>();
		public Opt<Sheet> currentSheet = Opt.none();
		public String name;

		public ScreenState(String name) {
			this.name = name;
		}
	}
}
